src/gui/util/qundostack.cpp

```diff
@@ -196,8 +196,13 @@ bool QUndoCommand::mergeWith(const QUndoCommand *command)
 
 void QUndoCommand::redo()
 {
-    for (int i = 0; i < d->child_list.size(); ++i)
+    for (int i = 0; i < d->child_list.size();) {
         d->child_list.at(i)->redo();
+        if (d->child_list.at(i)->isObsolete())
```

Comment:
(Now that ```d->child_list.at(i)``` is repeated, there's a case for saving it in a local variable to save duplication.)

Why do the deletion after redoing (or, below, undoing) if the operation is obsolete ?  Surely, if it makes no difference, it can simply be removed without {un,re}doing.

For that matter, why even delete it ?

Reply:
![image](https://github.com/user-attachments/assets/136093ff-131f-46a1-925f-fc42ecda0224)
![image](https://github.com/user-attachments/assets/96c1a8c6-1103-4f66-8597-fd1eb247c267)
why remove the command from undostack->d->command_list after undo if it was set obsolete.
but why not remove the command from undocommand->d->child_list after undo if it was set obsolete.

Reply:
That rationale would be worth stating in the commit message.
I also note that the quoted code checks whether the command is obsolete, skipping it if it already is, then checks again after undoing or redoing it (which may have marked it obsolete) prior to removing it.

Would it make sense to condition the undo or redo here, likewise, on not being already obsolete ?

Reply:
I also think it make no sense.

Reply:
If the operation is obsolete, it makes as much sense to skip it as to delete it.
As explained in the comment on the ```QUndoStack``` code, redoing or undoing may set the obsolete flag.  So conditioning the action on ```!isObsolete()``` and then retesting ```isObsolete()``` after the action to decide whether to remove the entry makes reasonable sense.
I'm not familiar enough with the undo-stack code to know whether that reasoning also applies in ```QUndoCommand```, but it at least seems as plausible as the case for removal of obsolete entries there.